app/server/lib/AttachmentStoreProvider.ts

```diff
@@ -0,0 +1,97 @@
+import { IAttachmentStore } from "app/server/lib/AttachmentStore";
+import log from 'app/server/lib/log';
+import { ICreateAttachmentStoreOptions } from "./ICreate";
+
+export type AttachmentStoreId = string
+
+/**
+ * Creates an {@link IAttachmentStore} from a given store id, if the Grist installation is configured with that store's
+ * unique id.
+ *
+ * Each store represents a specific location to store attachments at, for example a "/attachments" bucket on MinIO,
+ * or "/srv/grist/attachments" on the filesystem.
+ *
+ * Attachments in Grist Documents are accompanied by the id of the store they're in, allowing Grist to store/retrieve
+ * them as long as that store exists on the document's installation.
+ */
+export interface IAttachmentStoreProvider {
+  // Returns the store associated with the given id, returning null if no store with that id exists.
+  getStore(id: AttachmentStoreId): Promise<IAttachmentStore | null>
+
+  getAllStores(): Promise<IAttachmentStore[]>;
+
+  storeExists(id: AttachmentStoreId): Promise<boolean>;
+
+  listAllStoreIds(): AttachmentStoreId[];
+}
+
+export interface IAttachmentStoreSpecification {
+  name: string,
+  create: (storeId: string) => Promise<IAttachmentStore>,
+}
+
+interface IAttachmentStoreDetails {
+  id: string;
+  spec: IAttachmentStoreSpecification;
+}
+
+export class AttachmentStoreProvider implements IAttachmentStoreProvider {
+  private _storeDetailsById: { [storeId: string]: IAttachmentStoreDetails } = {};
+
+  constructor(
+    _backends: IAttachmentStoreSpecification[],
+    _installationUuid: string
+  ) {
+    // In the current setup, we automatically generate store IDs based on the installation ID.
+    // The installation ID is guaranteed to be unique, and we only allow one store of each backend type.
```

Comment:
`we only allow one store of each backend type` - I thought we agreed to move on from this as too restrictive? It makes no sense to me to allow say an s3 bucket and an azure bucket but then to kick up a fuss if the azure bucket is replaced with a distinct s3 bucket.

Reply:
I remember discussing permitting a small label that would take precedence over storeSpec.name if given.

Reply:
We've since discussed this, and will use a "JSON in environment variable" approach